query-planner-js/src/buildPlan.ts

```diff
@@ -1326,183 +1210,583 @@ class FetchDependencyGraph {
+          ? []
+          : this.toStringInternal(g, indent + "  ")))
+      .join('\n');
+  }
+}
+
+interface FetchGroupProcessor<G, P, F> {
+  onFetchGroup(group: FetchGroup, isRootGroup: boolean): G;
+  reduceParallel(values: G[]): P;
+  reduceSequence(values: (G | P)[]): G;
+  finalize(roots: G[], isParallel: boolean): F
+}
+
+export function computeQueryPlan(supergraphSchema: Schema, federatedQueryGraph: QueryGraph, operation: Operation): QueryPlan {
+  if (operation.rootKind === 'subscription') {
+    throw new GraphQLError(
+      'Query planning does not support subscriptions for now.',
+      [parse(operation.toString())],
+    );
+  }
+  // We expand all fragments. This might merge a number of common branches and save us
+  // some work, and we're going to expand everything during the algorithm anyway.
+  operation = operation.expandAllFragments();
+  operation = withoutIntrospection(operation);
+
+  debug.group(() => `Computing plan for\n${operation}`);
+  if (operation.selectionSet.isEmpty()) {
+    debug.groupEnd('Empty plan');
+    return { kind: 'QueryPlan' };
+  }
+
+  const root = federatedQueryGraph.root(operation.rootKind);
+  assert(root, () => `Shouldn't have a ${operation.rootKind} operation if the subgraphs don't have a ${operation.rootKind} root`);
+  const processor = fetchGroupToPlanProcessor(operation.variableDefinitions, operation.selectionSet.fragments, operation.name);
+  if (operation.rootKind === 'mutation') {
+    const dependencyGraphs = computeRootSerialDependencyGraph(supergraphSchema, operation, federatedQueryGraph, root);
+    const rootNode = processor.finalize(dependencyGraphs.flatMap(g => g.process(processor)), false);
+    debug.groupEnd('Mutation plan computed');
+    return { kind: 'QueryPlan', node: rootNode };
+  } else {
+    const dependencyGraph =  computeRootParallelDependencyGraph(supergraphSchema, operation, federatedQueryGraph, root);
+    const rootNode = processor.finalize(dependencyGraph.process(processor), true);
+    debug.groupEnd('Query plan computed');
+    return { kind: 'QueryPlan', node: rootNode };
+  }
+}
+
+function isIntrospectionSelection(selection: Selection): boolean {
+  return selection.kind == 'FieldSelection' && selection.element().definition.isIntrospectionField();
+}
+
+function mapOptionsToSelections<RV extends Vertex>(
+  selectionSet: SelectionSet,
+  options: SimultaneousPathsWithLazyIndirectPaths<RV>[]
+): [Selection, SimultaneousPathsWithLazyIndirectPaths<RV>[]][]  {
+  // We reverse the selections because we're going to pop from `openPaths` and this ensure we end up handling things in the query order.
+  return selectionSet.selections(true).map(node => [node, options]);
+}
+
+function possiblePlans(closedBranches: SimultaneousPaths<any>[][]): number {
+  let totalCombinations = 1;
+  for (let i = 0; i < closedBranches.length; ++i){
+    const eltSize = closedBranches[i].length;
+    if(!eltSize) {
+      totalCombinations = 0;
```

Comment:
Is this right? If any branch has a length of 0, we want the totalCombinations to be 0? If that's true, maybe just return 0 here?

Reply:
It's right. It's basically saying that if we find no path whatsoever for a queried field, then well, we cannot plan the overall query at all. Of course, in practice, this shouldn't happen if composition validation is not buggy and has run successfully, since it's the whole point of composition validation to ensure we don't get this during query planning.

Anyway, switched to `return 0` and added a comment.